crates/y-sweet/src/server.rs

```diff
@@ -101,38 +128,137 @@ impl Server {
         {
             let sync_kv = dwskv.sync_kv();
             let checkpoint_freq = self.checkpoint_freq;
+            let doc_id_ = doc_id.to_string();
             let doc_id = doc_id.to_string();
-            tokio::spawn(
-                async move {
-                    // TODO: expedite save on shutdown.
-                    let mut last_save = std::time::Instant::now();
-
-                    while let Some(()) = recv.recv().await {
-                        tracing::info!("Received dirty signal.");
-                        let now = std::time::Instant::now();
-                        if now - last_save < checkpoint_freq {
-                            let timeout = checkpoint_freq - (now - last_save);
-                            tracing::info!(?timeout, "Throttling.");
-                            tokio::time::sleep(timeout).await;
-                            tracing::info!("Done throttling.");
-                        }
-
-                        tracing::info!("Persisting.");
-                        sync_kv.persist().await.unwrap();
-                        last_save = std::time::Instant::now();
-                        tracing::info!("Done persisting.");
-                    }
+            let persistence_cancellation_token = CancellationToken::new();
+
+            // Spawn a task to save the document to the store when it changes.
+            self.doc_worker_tracker.spawn(
+                Self::doc_persistence_worker(
+                    recv,
+                    sync_kv,
+                    checkpoint_freq,
+                    doc_id,
+                    persistence_cancellation_token.clone(),
+                )
+                .instrument(span!(Level::INFO, "save_loop", doc_id=?doc_id_)),
+            );
 
-                    tracing::info!("Terminating loop.");
-                }
-                .instrument(span!(Level::INFO, "save_loop", doc_id=?doc_id)),
+            self.doc_worker_tracker.spawn(
+                Self::doc_gc_worker(
+                    self.docs.clone(),
+                    doc_id_.clone(),
+                    checkpoint_freq,
+                    persistence_cancellation_token,
+                    self.cancellation_token.clone(),
+                )
+                .instrument(span!(Level::INFO, "gc_loop", doc_id=?doc_id_)),
             );
         }
 
         self.docs.insert(doc_id.to_string(), dwskv);
         Ok(())
     }
 
+    async fn doc_gc_worker(
+        docs: Arc<DashMap<String, DocWithSyncKv>>,
+        doc_id: String,
+        checkpoint_freq: Duration,
+        persistence_cancellation_token: CancellationToken,
+        cancellation_token: CancellationToken,
+    ) {
+        let mut last_had_references = std::time::Instant::now();
```

Comment:
As I understand the logic here, we're GCing if the doc has not had references for two consecutive checks. If that's the case, I think it would be clearer to follow the logic here if it were just a `u8` called something like `checkpoints_without_refs` with these rules:

- it starts as `0`
- if a doc has exactly 1 strong reference, increment it by `1`
- if a doc has > 1 strong reference, reset it to `0`
- if it reaches `2`, GC the doc